pydantic/_internal/_typing_extra.py

```diff
@@ -221,12 +221,36 @@ def eval_type_lenient(value: Any, globalns: dict[str, Any] | None, localns: dict
         value = _make_forward_ref(value, is_argument=False, is_class=True)
 
     try:
-        return typing._eval_type(value, globalns, localns)  # type: ignore
+        return eval_type_backport(value, globalns, localns)
     except NameError:
         # the point of this function is to be tolerant to this case
         return value
 
 
+def is_unsupported_types_for_union_error(e: TypeError) -> bool:
+    return str(e).startswith('unsupported operand type(s) for |: ')
+
+
+def eval_type_backport(
+    value: Any, globalns: dict[str, Any] | None = None, localns: dict[str, Any] | None = None
+) -> Any:
+    """Like `typing._eval_type`, but falls back to the `eval_type_backport` package if it's
+    installed to let older Python versions use newer typing features.
+    Currently this just means that `X | Y` is converted to `Union[X, Y]` if `X | Y` is not supported.
+    This would also be the place to add support for `list[int]` instead of `List[int]` etc.
```

Comment:
I think it makes sense to support at least the most well known cases of `type`, `list`, `dict`, `set` when we release this.

Presumably there's a set list somewhere and implementing more things is a simple as a lookup dict?

I think supporting pipe unions but nothing else is confusing.